lisa/sut_orchestrator/azure/features.py

```diff
@@ -108,10 +121,126 @@ def _execute(self, wait: bool, operator: str, **kwargs: Any) -> None:
+                "resource_group_name", resource_group_name, "str"
+            ),
+            "resourceProviderNamespace": self._serialize.url(
+                "resource_provider_namespace", resource_provider_namespace, "str"
+            ),
+            "parentResourceType": self._serialize.url(
+                "parent_resource_type", parent_resource_type, "str", skip_quote=True
+            ),
+            "parentResource": self._serialize.url(
+                "parent_resource", parent_resource, "str"
+            ),
+            "serialPort": self._serialize.url("serial_port", serial_port, "str"),
+            "subscriptionId": self._serialize.url(
+                "self._config.subscription_id", self._config.subscription_id, "str"
+            ),
+        }
+        url = self._client.format_url(url, **path_format_arguments)
+
+        # Construct parameters
+        query_parameters = {}  # type: Dict[str, Any]
+        query_parameters["api-version"] = self._serialize.query(
+            "api_version", api_version, "str"
+        )
+
+        # Construct headers
+        header_parameters = {}  # type: Dict[str, Any]
+        header_parameters["Content-Type"] = self._serialize.header(
+            "content_type", content_type, "str"
+        )
+        header_parameters["Accept"] = self._serialize.header("accept", accept, "str")
+
+        request = self._client.post(url, query_parameters, header_parameters)
+        pipeline_response = self._client._pipeline.run(request, stream=False, **kwargs)
+        response = pipeline_response.http_response
+
+        if response.status_code not in [200]:
+            map_error(  # type: ignore
+                status_code=response.status_code, response=response, error_map=error_map
+            )
+            raise HttpResponseError(
+                response=response, error_format=ARMErrorFormat
+            )  # type: ignore
+
+        deserialized = self._deserialize("SerialPortConnectResult", pipeline_response)
+
+        if cls:
+            return cls(pipeline_response, deserialized, {})
+
+        return deserialized
+
+    connect.metadata = {  # type: ignore
+        "url": (
+            "/subscriptions/{subscriptionId}/"
+            "resourcegroups/{resourceGroupName}/"
+            "providers/{resourceProviderNamespace}/"
+            "{parentResourceType}/{parentResource}/"
+            "providers/Microsoft.SerialConsole/"
+            "serialPorts/{serialPort}/connect"
+        )
+    }
+
+
 class SerialConsole(AzureFeatureMixin, features.SerialConsole):
+    RESOURCE_PROVIDER_NAMESPACE = "Microsoft.Compute"
+    PARENT_RESOURCE_TYPE = "virtualMachines"
+    DEFAULT_SERIAL_PORT_ID = 0
+
     def _initialize(self, *args: Any, **kwargs: Any) -> None:
         super()._initialize(*args, **kwargs)
         self._initialize_information(self._node)
+        self._initialize_serial_console(id=self.DEFAULT_SERIAL_PORT_ID)
+
+    def run(self, cmd: str) -> str:
```

Comment:
It's not easy to use like this way, because most scenarios it needs to login, and the console may output other system logs unpredictable. So let's create a connection and return in this iteration. So it returns a writer and reader. The writer is used to send content to serial console, and the reader is used to read back.

Reply:
This was my initial thought as well. However, the websocket connection is very unreliable and each time we make a new connection, output buffer resets, and user would need to read everything from beginning. The things that worked for me is to create a websocket connection, flush out the current output, send command to the websocket and read back the output.  If we just return a reader and writer, all the connection management would become an issue for users.

Reply:
The input and output is not 1:1 mapping. We should assume the web socket stable at least when it's assigned. We can control underlying when to deliver a new connection or reuse existing one.